arch/arm/core/aarch64/cpu_idle.S

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright (c) 2019 Carlo Caione <ccaione@baylibre.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief ARM64 Cortex-A power management
+ */
+
+#include <toolchain.h>
+#include <linker/sections.h>
+
+#define DAIFSET_IRQ	((1) << 1)
+#define DAIF_IRQ	((1) << 7)
+
+GTEXT(arch_cpu_idle)
+SECTION_FUNC(TEXT, arch_cpu_idle)
+	msr	daifclr, #DAIFSET_IRQ
+	isb
+	wfi
```

Comment:
Shouldn't we unlock the IRQs before `wfi`?

Reply:
Not needed.  interrupt is 'taken' and wfi instruction will complete regardless of  PSTATE.I  but keeping it masked ensures a controlled exit. true for V7 cpsr.I . 
 Enabling it before wfi is rather dangerous as the wfi will complete only on next interrupt if any !

Reply:
> Enabling it before wfi is rather dangerous as the wfi will complete only on next interrupt if any !

For tickful mode, there is a constant supply of the periodic system timer interrupt, so we do not need to worry about this.

For tickless mode, `z_clock_announce`, called by the system timer ISR, will always configure the system timer to generate an interrupt after the calculated idle (shortest timeout) interval, after which the kernel will switch to the non-idle task; if a system timer interrupt occurs in `arch_cpu_idle` after enabling IRQ and before executing `wfi`, and the kernel returns to the idle thread to execute the `wfi` instruction thereafter, the system timer would have already been configured to generate an interrupt after the calculated idle interval; as for a non-system timer interrupt occurring in such a scenario, the previous system timer interrupt would have already configured the system timer, so this is not a problem.

It is true that `wfi` treats interrupts as a wake-up event regardless of whether interrupt is masked or not; the interrupt service routine, however, will not be executed after wake-up if interrupt is masked and, for a system timer interrupt, this means the timer driver will not be notified of the wake-up; hence, breaking the tickless kernel operation.

Reply:
I guess that the @stephanosio great explanation settles this.

Reply:
> this means the timer driver will not be notified of the wake-up; hence, breaking the tickless kernel operation
> We must un-mask IRQ before `wfi`

@stephanosio    I can only assume you missed to read my suggested code
(where we do unmask interrupt, but this time rightly after 'wfi'.).
 The suggested code gives a controlled exit from wfi instead of vectoring away on interrupt and comeback to wait for the same expecting there will be more anyways. In co-operative scheduling this may have a race condition or add a delay in yielding out of idle task.

Reply:
I gave this a look and it seems that indeed we need to have `wfi` before unmasking the interrupts. This seems confirmed by http://infocenter.arm.com/help/topic/com.arm.doc.prdc013242/Critical_Interrupt_Prioritization_PRD04_PRDC_013242_1_0.pdf and by the Linux code for `arch_cpu_idle()`.